modules/reindex/src/main/java/org/elasticsearch/index/reindex/remote/RemoteScrollableHitSource.java

```diff
@@ -0,0 +1,231 @@
+import org.elasticsearch.index.reindex.remote.RemoteRequestBuilders.InitialSearch;
+import org.elasticsearch.index.reindex.remote.RemoteRequestBuilders.Scroll;
+import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.io.BufferedInputStream;
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+
+import static java.util.Collections.emptyMap;
+import static org.elasticsearch.common.unit.TimeValue.timeValueMillis;
+import static org.elasticsearch.common.unit.TimeValue.timeValueNanos;
+import static org.elasticsearch.index.reindex.remote.RemoteResponseParsers.MAIN_ACTION_PARSER;
+import static org.elasticsearch.index.reindex.remote.RemoteResponseParsers.RESPONSE_PARSER;
+
+public class RemoteScrollableHitSource extends ScrollableHitSource {
+    private final AsyncClient client;
+    private final BytesReference query;
+    private final SearchRequest searchRequest;
+    Version remoteVersion;
+
+    public RemoteScrollableHitSource(ESLogger logger, BackoffPolicy backoffPolicy, ThreadPool threadPool, Runnable countSearchRetry,
+            Consumer<Throwable> fail, AsyncClient client, BytesReference query, SearchRequest searchRequest) {
+        super(logger, backoffPolicy, threadPool, countSearchRetry, fail);
+        this.query = query;
+        this.searchRequest = searchRequest;
+        this.client = client;
+    }
+
+    @Override
+    public void close() {
+        try {
+            client.close();
+        } catch (IOException e) {
+            fail.accept(new IOException("Couldn't close the remote connection.", e));
+        }
+    }
+
+    @Override
+    protected void doStart(Consumer<? super InternalResponse> onResponse) {
+        lookupRemoteVersion(version -> {
+            remoteVersion = version;
+            execute("POST", InitialSearch.path(searchRequest), InitialSearch.params(searchRequest, version),
+                    InitialSearch.entity(query), RESPONSE_PARSER, r -> onStartResponse(onResponse, r));
+        });
+    }
+
+    void lookupRemoteVersion(Consumer<Version> onVersion) {
+        execute("GET", "", emptyMap(), null, MAIN_ACTION_PARSER, onVersion);
+        
+    }
+
+    void onStartResponse(Consumer<? super InternalResponse> onResponse, InternalResponse response) {
+        if (Strings.hasLength(response.getScroll()) && response.getHits().isEmpty()) {
+            logger.debug("First response looks like a scan response. Jumping right to the second. scroll=[{}]", response.getScroll());
+            doStartNextScroll(response.getScroll(), timeValueMillis(0), onResponse);
+        } else {
+            onResponse.accept(response);
+        }
+    }
+
+    @Override
+    protected void doStartNextScroll(String scroll, TimeValue extraKeepAlive, Consumer<? super InternalResponse> onResponse) {
```

Comment:
It drives me bonkers that this is called "scroll" everywhere instead of "scrollId", but it's a matter of taste, no impetus to change it if you like it :)